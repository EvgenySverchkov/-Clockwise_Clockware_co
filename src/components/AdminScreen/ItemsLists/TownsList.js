import React from "react";
import { useDispatch, useSelector } from "react-redux";
import PropTypes from "prop-types";

import {
  deleteTownFromState,
  townsInit,
} from "../../../store/adminPanel/towns/actions";
import {changeListIsLoad} from "../../../store/adminPanel/services/actions";
import deleteDataFromServer from "../services/deleteDataFromServer";

import List from "../List";

import { SERVERDOMAIN } from "../../../services/serverUrls";

const TownsList = ({ history }) => {
  const state = useSelector(state=>{
    return {
      townsArr: state.town_reduser.towns,
      listIsLoad: state.main_adminPanel_reduser.listIsLoad
    }
  });
  const dispatch = useDispatch();

  function getTownsFromServerToState() {
    const headers = {
      Authorization: "Bearer " + sessionStorage.getItem("token") || "",
      "Content-Type": "application/json",
    };
    dispatch(changeListIsLoad(true))
    fetch(`${SERVERDOMAIN}/towns`, { headers })
      .then((json) => json.json())
      .then((data) => {
        dispatch(changeListIsLoad(false))
        dispatch(townsInit(data))
      })
      .catch(err=>{
        dispatch(changeListIsLoad(false));
        alert(err)
      });
  }
  function deleteTownById(townId) {
    deleteDataFromServer(`${SERVERDOMAIN}/towns/delete/${townId}`)
      .then((data) => {
        if (data.success) {
          dispatch(deleteTownFromState(data.payload));
          alert(data.msg);
        } else {
          alert(data.msg);
        }
      })
      .catch((err) => alert(err));
  }
  return (
    <List
      dataArr={state.townsArr}
      deleteAction={deleteTownById}
      mainRows={["name", "id"]}
      getData={getTownsFromServerToState}
      history={history}
    />
  );
};

TownsList.propTypes = {
  history: PropTypes.object,
};

export default TownsList;
