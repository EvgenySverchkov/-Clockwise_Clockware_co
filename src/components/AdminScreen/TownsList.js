import React from 'react';
import {connect} from "react-redux";

import List from "./List";
import ListItem from "./ListItem";

import {deleteTown} from "../../store/adminPanel/actions";
import deleteBtnHoc from "../../hocs/withDeleteBtn";

function TownsList({townsArr, deleteTown}){
  function deleteTownById(townId){
    fetch(`https://clockwiseserver.herokuapp.com/delete_town/${townId}`, {
      method: "delete"
    }).then(data=>data.json())
    .then(data=>deleteTown(data))
  }
  let ListItemWithDelete = deleteBtnHoc(ListItem, deleteTownById);
  return <List ListItem = {ListItemWithDelete}
               dataArr={townsArr}
               style = {style}/>
}
let style = {
  width: '20%',
  margin: '0 auto'
}
export default connect(null, {deleteTown})(TownsList);
