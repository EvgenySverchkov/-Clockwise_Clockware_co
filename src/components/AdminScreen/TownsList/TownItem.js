import React from 'react';
export default ({name, id})=>{
  return (
    <>
      <th scope="row">{id}</th>
      <td>{name}</td>
    </>
  )
}
