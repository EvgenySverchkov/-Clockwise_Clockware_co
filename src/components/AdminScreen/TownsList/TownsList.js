import React from 'react';
import {connect} from "react-redux";
import TownItem from "./TownItem";
import {deleteTown} from "../../../store/adminPanel/actions";
import withDeleteBtn from "../../../hocs/withDeleteBtn";


function TownsList({townsArr, deleteTown}){
  let NewComp = withDeleteBtn(TownItem, deleteTown);
  return (
    <ul>
      {townsArr.map((item)=><NewComp key={item} item={item} by={item}/>)}
    </ul>
  );
}
export default connect(null, {deleteTown})(TownsList);
