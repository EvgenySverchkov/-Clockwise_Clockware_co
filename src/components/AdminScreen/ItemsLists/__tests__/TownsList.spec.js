import React from "react";
import {create, act} from "react-test-renderer";
import {Provider} from "react-redux";
import mockStoreConfig from "redux-mock-store";
import {MemoryRouter} from "react-router-dom";
   
import TownsList from "../TownsList";

const mockStore = mockStoreConfig();

describe("Test of <TownsList />", ()=>{
    global.fetch = jest.fn(()=>Promise.resolve({json: jest.fn(()=>Promise.resolve())}));
    const store = mockStore({
        townReduser: {
            towns: [{
                id: 1,
                name: "Name"
            }]
        },
        adminModalWindows: {
            listIsLoad: false
        }
    });
    const mockProps = {
        history: {
            location: "/"
        }
    }
    it("When list is already loaded", async()=>{
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store={store}>
                    <MemoryRouter initialEntries={["/"]}>
                        <TownsList {...mockProps}/>
                    </MemoryRouter>
                </Provider>
            );
        })
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("When list is loading", async()=>{
        const store = mockStore({townReduser:{towns: []}, adminModalWindows: {listIsLoad: true}});
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store={store}>
                    <MemoryRouter initialEntries={["/"]}>
                        <TownsList {...mockProps}/>
                    </MemoryRouter>
                </Provider>
            );
        })
        expect(component.toJSON()).toMatchSnapshot();
    });
});