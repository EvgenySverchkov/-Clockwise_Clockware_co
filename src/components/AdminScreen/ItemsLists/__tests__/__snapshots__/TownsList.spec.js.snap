// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <TownsList /> When list is already loaded 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Id
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style={
            Object {
              "wordWrap": "break-word",
            }
          }
        >
          Name
        </td>
        <td
          style={
            Object {
              "wordWrap": "break-word",
            }
          }
        >
          1
        </td>
        <td
          className="text-center"
        >
          <div
            className="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              className="btn btn-secondary dropdown-toggle"
              data-toggle="dropdown"
              id="dropdownMenuButton"
              type="button"
            >
              Options
            </button>
            <div
              aria-labelledby="dropdownMenuLink"
              className="dropdown-menu"
            >
              <button
                className="dropdown-item deleteItemBtn"
                onClick={[Function]}
              >
                Delete
              </button>
              <a
                className="dropdown-item"
                href="/1"
                onClick={[Function]}
              >
                Edit
              </a>
              <button
                className="dropdown-item"
                data-target="#fullInfoModal"
                data-toggle="modal"
                onClick={[Function]}
                type="button"
              >
                Show full info
              </button>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Test of <TownsList /> When list is loading 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr>
        <th>
          List is empty
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
